### ISSUE TITLE
Make $route.hash mutable

### ISSUE THREAD

**Issue by woozyking** [vuejs/vue-router#1223]
Apologies if this has been raised up before (in which case I have failed at searching through the closed issues).

Is there any obvious reasons that we do not want to provide a mutable `$route.hash` property?

One thing I can think of is `history.pushState` compatibility which we may be able to just use a simple `if..else` to fallback to `location.hash`:

```js
if (window.history.pushState) {
  window.history.pushState(null, null, '#hash|values')
} else {
  window.location.hash = '#hash|values'
}
```

**Comment by LinusBorg** [comment]
> Is there any obvious reasons that we do not want to provide a mutable $route.hash property?

Is there a specific use case for which it should be?

the whole $route object is generally immutable - a new one gets created for every navigation. I don't see what's special about the hash prop on the $route object.

**Comment by woozyking** [comment]
This is a fresh use case I've just encountered -- imagine a view with such component http://element.eleme.io/#/en-US/component/collapse where a user would be able to share a link of the view with the states of the collapses (on/off) indicated and synced through URL hash.

For example `http://example.com/collapses#consistency|efficiency` would open the view with `consistency` and `efficiency` collapse sections opened and the others closed.

It's probably a minor use case, and pretty simple to achieve by manually doing what I proposed about `pushState`/`location.hash`. But if there's no anti-use case, I would think this is a really good convenience feature to make `$route.hash` mutable.
